test/integrations/hyperdrive/PriceDiscovery.t.sol

```diff
@@ -0,0 +1,569 @@
+        hyperdrive.initialize(
+            initialContribution,
+            fixedAPR,
+            IHyperdrive.Options({
+                destination: alice,
+                asBase: true,
+                extraData: new bytes(0)
+            })
+        );
+
+        // Alice opens a max long.
+        openLong(alice, hyperdrive.calculateMaxLong());
+        assertApproxEqAbs(hyperdrive.calculateSpotAPR(), 0, 100 wei);
+
+        // Alice adds liquidity again
+        addLiquidity(alice, addLiquidityContribution);
+
+        uint256 sharePrice = hyperdrive.getPoolInfo().vaultSharePrice;
+        uint256 shareReserves = hyperdrive.getPoolInfo().shareReserves;
+        uint256 longExposure = hyperdrive.getPoolInfo().longExposure;
+        int256 solvency = int256(shareReserves.mulDown(sharePrice)) -
+            int256(longExposure) -
+            int256(2 * minimumShareReserves.mulDown(sharePrice));
+        assertTrue(solvency > 0);
+    }
+
+    function test_solvency_cross_checkpoint_long_short(
+        uint256 fixedAPR,
+        uint256 initialContribution,
+        uint256 addLiquidityContribution,
+        bool longFirst
+    ) external {
+        uint256 minimumShareReserves = 10e18;
+
+        // Normalize the fuzzing parameters to a reasonable range.
+        fixedAPR = fixedAPR.normalizeToRange(0.01e18, 2e18);
+        uint256 timeStretchAPR = fixedAPR / 4;
+        timeStretchAPR = timeStretchAPR.max(0.01e18);
+        timeStretchAPR = timeStretchAPR.min(0.3e18);
+        initialContribution = initialContribution.normalizeToRange(
+            10_000e18,
+            100_000_000e18
+        );
+        addLiquidityContribution = addLiquidityContribution.normalizeToRange(
+            1e18,
+            100_000_000e18
+        );
+
+        // Configure the pool.
+        IHyperdrive.PoolConfig memory config = testConfig(
+            timeStretchAPR,
+            POSITION_DURATION
+        );
+        config.circuitBreakerDelta = type(uint128).max;
+        config.minimumShareReserves = minimumShareReserves;
+        deploy(alice, config);
+        vm.stopPrank();
+        vm.startPrank(alice);
+        baseToken.mint(initialContribution);
+        baseToken.approve(address(hyperdrive), initialContribution);
+        hyperdrive.initialize(
+            initialContribution,
+            fixedAPR,
+            IHyperdrive.Options({
+                destination: alice,
+                asBase: true,
+                extraData: new bytes(0)
+            })
+        );
+
+        // Alice opens a max long or a max short.
+        if (longFirst) {
+            openLong(alice, hyperdrive.calculateMaxLong());
+        } else {
+            openShort(alice, hyperdrive.calculateMaxShort());
+        }
+
+        // The checkpoint passes.
+        advanceTime(CHECKPOINT_DURATION, 0);
+
+        // Alice opens a max long or a max short.
+        if (longFirst) {
+            openShort(alice, hyperdrive.calculateMaxShort());
```

Comment:
IMO, it would be good to test with amounts other than the max long and short. I’m not saying this isn’t sufficient, but without looking at this closer (which I will do tomorrow on the flights), I still have a few cases that I’m unsure about.